clang/test/CodeGenSYCL/spir-calling-conv.cpp

```diff
@@ -9,7 +9,9 @@ int main() {
 
```

Comment:
The IR for this code has simplified from -

      ; Function Attrs: norecurse nounwind
      define spir_kernel void @_ZTSZ4mainE15kernel_function() #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !4 
      !kernel_arg_type !4 !kernel_arg_base_type !4 !kernel_arg_type_qual !4 {
      entry:
          %0 = alloca %"class._ZTSZ4mainE3$_0.anon", align 1
          %1 = bitcast %"class._ZTSZ4mainE3$_0.anon"* %0 to i8*
          call void @llvm.lifetime.start.p0i8(i64 1, i8* %1) #3
          %2 = addrspacecast %"class._ZTSZ4mainE3$_0.anon"* %0 to %"class._ZTSZ4mainE3$_0.anon" addrspace(4)*
          call spir_func void @"_ZZ4mainENK3$_0clEv"(%"class._ZTSZ4mainE3$_0.anon" addrspace(4)* %2)
          %3 = bitcast %"class._ZTSZ4mainE3$_0.anon"* %0 to i8*
          call void @llvm.lifetime.end.p0i8(i64 1, i8* %3) #3
         ret void
     }



to
         ; Function Attrs: norecurse nounwind
         define spir_kernel void @_ZTSZ4mainE15kernel_function(%"class._ZTSZ4mainE3$_0.anon"* 
         byval(%"class._ZTSZ4mainE3$_0.anon") align 1 %_arg_kernelObject) #0 !kernel_arg_addr_space !4 
         !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_base_type !6 !kernel_arg_type_qual !7 {
        entry:
             %0 = addrspacecast %"class._ZTSZ4mainE3$_0.anon"* %_arg_kernelObject to 
             %"class._ZTSZ4mainE3$_0.anon" addrspace(4)*
            call spir_func void @"_ZZ4mainENK3$_0clEv"(%"class._ZTSZ4mainE3$_0.anon" addrspace(4)* %0)
             ret void
        }   

@Fznamznon Are those lifetime markers needed? I'm not sure what the intrinsic really does and if it's required.

Reply:
Lifetime markers are needed to specify start/stop of memory objects life time (oops I'm a captain obvious here). Lifetime markers define kinda frame outside of which the memory object is dead and has undefined value (see https://llvm.org/docs/LangRef.html#id2407 , it can explain better).
In the first LLVM IR snippet I see lifetime markers only for local kernel object clone. If we are going to remove local kernel object clone, then no need to keep lifetime markers for it :)
I expect exactly the same specification in resulting LLVM IR which I see in your second snippet, so it is nice progress.